File: frontend/src/api/operations.ts
import { UseQueryResult } from "react-query";
import { User } from "./user";
import useUser from "./user";

const Operations = {
	page_debug: "page_debug",
	page_users: "page_users",
	page_staff: "page_staff",
	page_staff_quest: "page_staff_quest",
	page_staff_store: "page_staff_store",
	page_stats: "page_stats",
	site_config: "site_config",
	promote_staff: "promote_staff",
	promote_manager: "promote_manager",
	manage_store: "manage_store",
	manage_quest: "manage_quest",
	manage_games: "manage_games",
	send_any: "send_any",
}

export type Operation = keyof typeof Operations;

export default function hasPermission(user: UseQueryResult<User, unknown>, operation: Operation)
{
	return !!(user.data?.auth && user.data?.operations.includes(operation));
}

export function useHasPermission(operation: Operation)
{
	const user = useUser();
	return hasPermission(user, operation);
}

File: backend/data/_operations.py
from bafser import OperationsBase


class Operations(OperationsBase):
    page_dev = ("page_dev", "Страница отладки")
    page_staff = ("page_staff", "Страница для персонала")
    page_staff_quest = ("page_staff_quest", "Страница кодов квестов")
    page_staff_store = ("page_staff_store", "Страница кодов товаров")
    page_stats = ("page_stats", "Страница статистики")

    promote_staff = ("promote_staff", "Повысить до волонтёра")
    promote_manager = ("promote_manager", "Повысить до управляющего")

    manage_store = ("manage_store", "Управление товарами")
    manage_quest = ("manage_quest", "Управление квестами")
    manage_games = ("manage_games", "Управление играми")

    add_any_image = ("add_any_image", "Добавление картинок")
    send_any = ("send_any", "Любые пересылки")
    site_config = ("site_config", "Настройка сайта")
